### ISSUE TITLE
Variety of automatic Benchmarks after installation, and potentially on-going QC

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#51]
We should do a series of benchmarks after installation, log the results.

Yabs.sh and network-speed.xyz should be ran.
a cronjob perhaps on the background which runs these tests in 60 minute intervals again and again until there is users in the server.

Log all the results, and have them so that users can view them. stored for long term with datetime, how long it took to run etc.

We should also run fio test designed to trash the I/O subsystem a lot like a seedbox would; 95% read, 5% write, random I/O 512-1024KiB request sizes, queue depths of 32 to 100 or so. Perhaps randomize which settings to test.
Further; sometimes FIO should be run so that at some random point a heavy I/O sequential process starts, ie. pigz/tar to compress set of data, or a file streamed to /dev/null @ 50Mbits, file copied over to /dev/null @ 115MiB/s, huge sequential filewrite at max speed single thread (ie. dd dev null to some file bs=1M size 50gigs)  etc. typical seedbox type loads.
Perhaps even run multiple FIO instances at the same time despite it is good to run. Type of test needs to be logged and be programmatically accessible so analytics can be done.

All of the tests on different files etc. so they can be used separately later on.

We should probably also run FIO in production with users every few weeks or so to make sure performance is what it ought to be, and if there's headroom so we can use that data for provisioning to disable new users for the server if performance is not high enough.

Save all results programmatically as well, in case we need them later on.

**Comment by MagnaCapax** [comment]
Actually the fio/I/O tests should also take iostat numbers during the test, this gives the most programmatically accessible data and is not dependant on fio reports alone.
FIO instances should be multitude started at random times, random settings, random durations, random data sets etc. to simulate real world utilization the best.

Might be good idea to do 2 minute test weekly, and 10minute test monthly or something like that. Can't be any longer than that, as to not cause issues for user.

GUI should show when a performance test is being run, just like other potential performance warnings mentioned in some issues.

**Comment by TankTetrarch** [comment]
test services instantiated, verify credentials against services, port forwarding, metrify delays, disk thrash rates, disclose throttles in pretty graphs

**Comment by MagnaCapax** [comment]
Related performance-metrics work:
- #45 (long-term storage I/O stats)
- #32 (end-user IOPS graphs)
- #169 (ioping availability for disk latency metrics)
- #42 / #112 / #120 (observability transport + correlation/profiling)

— Sampsa Pellervoinen 🌱

**Comment by MagnaCapax** [comment]
Related: #45 (local long-term IO performance stats / rollups). Benchmarks/QC outputs could feed or validate the same performance-metrics pipeline.

— Sampsa Pellervoinen 🌱

**Comment by MagnaCapax** [comment]
## Investigation Report

**Issue:** #51 — Variety of automatic Benchmarks after installation, and potentially on-going QC
**Investigated:** 2026-06-03
**Status:** PARTIAL

### Summary
This 2024 issue bundles six distinct asks. The codebase has quietly overtaken most of them: the storage-benchmark engine (`scripts/lib/storageBenchmark.php`) already implements the exact fio profile requested here, and a passive ongoing-QC pipeline already runs via cron. The one genuinely dangerous ask — running aggressive fio *in production with live users* and auto-gating provisioning on it — should not be built: it reproduces a known multi-VM outage and measures the wrong thing. The real provisioning-performance need behind it is better served by the passive-metrics work already in flight (#45/#501). Recommend **SPLIT_FIRST**.

### What The Issue Asks For
1. Post-install benchmarks (`yabs.sh` + a network speed test), logged.
2. A cronjob running benchmarks every 60 min "until there are users in the server".
3. Long-term, user-viewable result storage (datetime, duration).
4. fio I/O stress: 95% read / 5% write, random 512–1024 KiB, iodepth 32–100, randomized settings, mixed sequential-load injection (pigz/tar, stream to /dev/null, dd), multiple concurrent fio instances; capture iostat during the run.
5. **Active fio in production every few weeks** to verify performance, and **auto-disable new users** on a server when performance is too low (a provisioning gate).
6. GUI indicator while a perf test runs; ~2-min weekly / ~10-min monthly cadence.

### Implementation Status
Already implemented (verified by reading source):
- **`scripts/lib/storageBenchmark.php`** (+ `scripts/lib/storageBenchmark/report.php`, `scripts/util/storageBenchmark.php`) is a non-destructive fio wrapper. Its job `randmix-large-95r5w` is `rw=randrw rwmixread=95 bssplit=4k/2:64k/3:128k/5:256k/10:512k/20:768k/25:1024k/35 iodepth=32 numjobs=4 direct=1` — **literally ask #4's single-job profile** — alongside randread-1M, randread-4k, randwrite-4k, seqread-1M.
- Per-device tests (`--devices`): dd seqread + fio randread per physical device, with **cross-device peer-comparison warnings** (a built-in outlier/regression detector).
- iostat util + ioping captured as a **preflight idle gate** (`--require-idle`, `--idle-latency-ms` default 100, `--idle-util` default 85).
- **JSONL log** `/var/log/pmss/benchmark-storage.jsonl` (run_id, timestamp, params, metrics: read/write MB/s, IOPS, p95) — programmatically accessible; `--show-last` renders it. Path/input hardening + a dedicated security test exist.
- **Passive ongoing QC already runs via `etc/seedbox/config/root.cron`:** `diskIostat.php` (5 min), `systemStatsLog.php` (5 min; ioping + multi-timescale PSI), `cpuStat.php`, `resourceLog.php`, `storageHealthSnapshot.php` (SMART, 2×/day), `processSnapshot.php`, MD `checkarray` (quarterly). iostat-history persistence landed via #581; device filter via #600.

### What's Missing
- No install-time trigger and no "until there are users" pre-user loop (`scripts/lib/user/selection.php` exists to reuse for the no-users check).
- `yabs.sh` and a network throughput test are absent (the motd `%NETWORK_SPEED%` is NIC link speed, not a throughput test).
- No GUI integration: `etc/skel/www/welcome.php` has none; no rollup-JSON / stats-page graph (that residue is tracked in #45).
- The benchmark runs a **fixed job list sequentially** — no concurrent fio instances, no randomized settings/durations/datasets, no mixed sequential-load injection.
- No active-production-fio QC and no auto-disable-new-users provisioning gate.

### Mission Alignment
- **Reliability over heroics (#7) / business longevity (#4):** the existing tool + passive cron serve QC without risking customer service. Ask #5 (active fio in production) directly endangers reliability — see Related Incidents.
- **Investigate then act (#8):** the right provisioning signal is observed contended load (passive PSI/serviceTime/iostat), not a synthetic idle-time peak. Building the dangerous mechanism would be acting before understanding.
- **Value for money (#5):** reuse of the already-built benchmark + observability stack beats re-implementing it.

### Devil's Advocate
*Why not build #51 as written?* Most of it already exists, the dangerous part should be refused, and `yabs.sh` (a third-party remote-script-piped-to-shell installer) adds supply-chain surface and duplicates the disk benchmark we already have. *Rebuttal for the part worth building:* a one-time **post-install idle baseline** is genuinely useful (acceptance test + capacity baseline + per-device outlier catch) and is safe precisely because there are no users yet — so a small, bounded slice does survive the cut.

### Anti-Pattern Hunt
- REGEX_FOR_SEMANTICS: PASS — no semantic classification involved; benchmark metrics are numeric.
- PREMATURE_OPTIMIZATION: PASS — recommendation is to NOT build the speculative production-QC/provisioning-gate machinery; the small slice (Exists: storageBenchmark; Tested: yes; Useful: install baseline; Too-expensive-now: the production-fio part) is justified by a real need.
- FALSE_COMPLETION_CLAIM: PASS — status PARTIAL; each "done" claim cites a file; gaps enumerated.
- OPERATOR_HANDOFF_ANTIPATTERN: PASS — primary recommendation is SPLIT_FIRST with concrete per-slice verdicts, not a deferral to the operator.
- FABRICATED_AUTHORITY_OR_PATH: PASS — every cited path read this run (`storageBenchmark.php`, `root.cron`, `welcome.php`, `docs/adr/`, `user/selection.php`).
- FAIL_OPEN_SAFETY_BEHAVIOR: PASS — the recommended slice enforces `--require-idle` (fail-closed: abort if busy), and explicitly refuses the run-while-busy path.
- DRY_RUN_THEATER: PASS — not applicable; investigation read real source + real incident records.
- INVESTIGATION_LOOP_THEATER: PASS — first investigation of #51; produces an actionable decomposition, not another "needs more research".

### Related Incidents & Data
- **Multi-VM host outage (2026-03-29):** a single user's aggressive `fio --direct=1` on a multi-tenant VM host triggered a kernel soft-lockup (load ~666), froze every VM on the host for **17+ hours**, and required a hardware power cycle. The benchmark's default jobs use `direct=1 iodepth=32 numjobs=4` — the same shape. This is safe on an empty/pre-user server and an incident reproducer on a populated one. It is the central reason to refuse ask #5.
- **Provisioning-performance gap (internal analysis, 2026-05):** capacity planning is performance-bound, not space-bound — i.e. the need behind ask #5 is real — but the indicated fixes are all *passive* metric improvements, not active benchmarking.
- Frequency: #51 has sat open ~2 years; no operational pressure for the active-fio part, while the passive-metrics work (#45/#501) is actively progressing.

### Cross-Issue Analysis
- **#45** (local long-term I/O performance stats / rollups) — OPEN/PARTIAL; the passive backbone this issue's outputs would feed. Persistence landed (#581/#600); rollup + stats-UI graph still missing. **DEPENDS / partial MERGE** (the GUI + result-graph portion of #51 belongs here).
- **#501** (9 missing per-user/VM metrics + iostat persistence; "closes #45") — OPEN, reworked, broader. The performance-aware-provisioning signal belongs to this metrics line, not to active fio.
- **#121** (cgroup IO/burst/net/NOFILE extensions) — OPEN; the in-guest I/O containment that makes *any* benchmarking safer; closed **#273** wired io.max. Landing-order: containment first, then anything that intentionally generates load.
- No file-level edit conflict with other open issues (no other open issue touches `storageBenchmark.php`).

### Dependencies & Blockers
- Slice 1 (post-install baseline) has no blockers — the tool exists; only an install/cron hook + a no-users guard are needed.
- The GUI/result-graph portion depends on #45's rollup/stats work.
- The provisioning-gate portion depends on the passive-metrics + capacity-planning line and lives outside this repo — out of scope for #51.

### Reuse, Refactor & Test Strategy
- **Existing reuse points:** `scripts/lib/storageBenchmark.php` (engine + idle gate + JSONL), `scripts/lib/user/selection.php` (no-users detection), `etc/seedbox/config/root.cron` (scheduling slot), the passive cron collectors above.
- **Refactor opportunities:** add a thin `--require-idle`-enforced invocation hook rather than new benchmark logic (DRY); fold any result display into the existing stats page rather than a new GUI surface (separation of concerns: collection vs presentation already split between `storageBenchmark.php` and `report.php`).
- **Test categories proposed:** hermetic unit for the no-users gate decision; reuse the existing `storageBench*Test.php` pattern for any new option; regression test that the install hook refuses to run when users exist / system is busy (guards against re-triggering the 2026-03-29 failure mode); boundary test for "exactly one user present".
- **Minimum-code-for-correctness:** an install-completion call to the existing tool with `--require-idle`, plus a user-count guard. No new abstraction, no configurability without a second caller.
- **Mission/business success signal:** a captured per-server idle baseline in `benchmark-storage.jsonl` at provisioning time (mission #8 investigate-then-act: acceptance evidence before a server takes customers); business signal: catches dud disks/arrays *before* a paying customer lands on them (retention + avoided refund/migration cost).

### Deep Research Gate
SKIPPED: internal-evidence-sufficient — the benchmark engine, cron pipeline, the fio-saturation incident analysis, and the provisioning-performance analysis are all already documented internally and were read this run; no novel domain, no industry comparison required, and #51 has not been investigated before.

### Architectural Minimization
**Step A — dumbest requirement:** "run fio in production every few weeks and auto-disable new users on low performance." Author = the operator, **in 2024** — before passive PSI/ioping/iostat collection existed and before the 2026 fio-saturation incident. The underlying constraint it serves (provisioning must respect I/O headroom) is real and still current; the *active-fio mechanism* is an inherited 2024 assumption that newer passive instrumentation made both unnecessary and unsafe.

**Step B — alternatives:**
- **DELETE:** (1) Delete ask #5's active-production-fio — the headroom signal is provided by the passive cron metrics (PSI/serviceTime/iostat) already collected; this is the **preferred path** for the production-QC portion. (2) Delete `yabs.sh`/network-test — the disk portion duplicates `storageBenchmark.php`, throughput is already in passive traffic logging, and a remote-script-piped-to-shell installer conflicts with PMSS's shell-execution-safety posture.
- **MERGE:** (1) Merge #51's result-GUI into #45's stats-page/rollup work (shared consumer of perf data). (2) Merge the performance-provisioning gate into the (out-of-repo) capacity-planning line.
- **REFACTOR:** (1) Refactor the install path to call the existing `storageBenchmark.php` once at completion (`--require-idle`). (2) Refactor `scripts/lib/user/selection.php` usage into a no-users guard for an optional pre-user loop.
- **NULL:** If nothing is done — no automatic fresh-server I/O baseline is captured; operators/agents run `storageBenchmark.php` manually during investigations, and passive QC + capacity planning continue unaffected. Concrete consequence: per-server acceptance/regression comparison stays manual. Modest impact.

**Step C — minimal design proposal:** Add an install-completion (and optional pre-user-loop) invocation of the existing `storageBenchmark.php` with `--require-idle` enforced and a user-count guard from `scripts/lib/user/selection.php`; results already land in `benchmark-storage.jsonl`. Files: a small hook in the install/update completion path + one cron-loop guard (~30–60 LOC total). Stripped vs the issue: active production fio, the auto-disable gate, yabs/network test, and a bespoke GUI — each stripped because it is dangerous, out-of-repo, redundant, or belongs to #45.

### ADR Adherence
NO_RELEVANT_ADRS: searched `docs/adr/`, listed 24 ADRs (0001–0024); none governs storage benchmarking or performance QC. The recommendation is consistent with the observability-oriented ADRs (0006 update/observability) and does not conflict with the cgroup ADRs (0003 v2 adoption, 0019 v1 pin) or 0023 (cron-service TasksMax guard) — the proposed slice adds one idle-gated invocation, not a new resource-control mechanism.

### Final AQFM Sweep
Question asked (verbatim):
> "Issue #51 (MagnaCapax/PMSS): Variety of automatic Benchmarks after installation and ongoing QC. Draft conclusion: PMSS already implements the core fio storage benchmark (storageBenchmark.php with 95r5w bssplit-to-1024k iodepth32 numjobs4, iostat+ioping+JSONL+per-device), and passive ongoing QC runs via cron (diskIostat, systemStatsLog with ioping/PSI, storageHealthSnapshot). Missing: install-time trigger, pre-user loop, yabs/network test, GUI, and an active-production-fio QC + auto-disable-new-users gate. Draft recommendation: SPLIT_FIRST — implement small post-install idle benchmark slice, SHOULD-NOT-IMPLEMENT the active production fio QC (caused a 17h 23-VM outage 2026-03-29, redundant with passive metrics), drop yabs, merge GUI into #45/#501. What contradicting evidence, missed precedent, prior operator decision, or relevant research could revise this draft?"

Sources returned: the cached copy of issue #51 only (synthesis was retrieval-thin — it did not surface the incident or metrics records, which were read directly elsewhere in this investigation).
Verdict: **HELD** — no contradicting evidence surfaced. The substantive refinement (below) came from the companion self-challenge searches, not from this synthesis call.

### Adversarial Self-Challenge

**Strongest finding 1**: `storageBenchmark.php` already implements ask #4's fio profile.
- Disproving search: re-read the job definitions and the run loop for concurrency/randomization. → Found a real limitation: jobs run **sequentially from a fixed list**; no multiple concurrent fio instances, no randomized settings/durations, no mixed sequential-load injection (the only randomness is the dd read offset).
- Outcome: **revised** — the *single-job profile* matches exactly, but the concurrency/randomization/mixed-load orchestration in ask #4 is genuinely unimplemented (now reflected in "What's Missing").

**Strongest finding 2**: running aggressive fio in production is dangerous (cite the outage).
- Disproving search 1: incident record + its adversarial review. → Confirmed: documented 17h/all-VMs outage; operator correction notes dedicated per-VM arrays (so the cascade mechanism is host CPU/controller saturation), but the outage and its trigger stand.
- Disproving search 2 (broader): searched for any record advocating periodic active benchmarking as desirable. → Found the provisioning-performance analysis instead, which points to *passive* fixes, not active fio.
- Outcome: **held**, and strengthened — the danger is real and the alternative is passive.

**Least-evidenced finding** (lowest citation count): "active production fio is redundant with passive metrics."
- Disproving searches: the provisioning-performance analysis + the fleet I/O-observability records. → These initially looked like they might *support* active benchmarking (a real capacity/performance gap exists). On reading, every indicated remedy is passive (fix the dead utilization metric, ship latency probes centrally, storage-class-aware thresholds, an I/O-saturation gate).
- Outcome: **revised/sharpened** — not "redundant" flatly, but "answers the wrong question": idle-time peak ≠ contended real-load performance, and the provisioning question is the latter, which passive metrics already measure. The *need* is real; the *mechanism* is wrong. This is now the core of the recommendation.

### Recommendation
**SPLIT_FIRST**
- **Slice 1 — IMPLEMENT (small, safe):** post-install idle baseline — invoke the existing `storageBenchmark.php` at install completion with `--require-idle` enforced; optional pre-user repeat loop guarded by a no-users check. Captures a one-time peak-capacity baseline + per-device outlier check on the empty server.
- **Slice 2 — SHOULD-NOT-IMPLEMENT:** active fio in production + auto-disable-new-users gate. The real provisioning-performance need is served by the passive-metrics line (#45/#501) and an I/O-saturation gate in the capacity-planning system, not by synthetic load on live customers.
- **Slice 3 — SHOULD-NOT-IMPLEMENT / drop:** `yabs.sh` + network throughput test (third-party remote-script-piped-to-shell surface; disk portion duplicates `storageBenchmark.php`; throughput already in passive traffic logging).
- **Slice 4 — MERGE into #45:** result GUI / stats-page graph for the benchmark JSONL.

Reasoning: most of this 2024 issue is already shipped; the unshipped remainder is one safe small slice plus several parts that should be refused or routed to their proper home. SPLIT_FIRST keeps the issue open for operator visibility on the decomposition rather than auto-closing a partly-valid request.

### Auto-Close Decision
AUTO_CLOSED_BLOCKED: trigger-not-met; reason=recommendation-is-IMPLEMENT/NEEDS_RESEARCH/NEEDS_OPERATOR_INPUT/SPLIT_FIRST

### Additive Recommendation Flags
- **RECOMMEND_ADR** — Draft title: *"Performance QC and provisioning gating are passive-metric-driven; active fio benchmarks run only on idle/pre-user servers, never in production with live users."* Context: PMSS already collects passive I/O telemetry (PSI/ioping/iostat) on a cron cadence and ships a non-destructive `storageBenchmark.php`. A 2026 incident showed aggressive fio on a populated multi-tenant host causes multi-hour outages. Options considered: (a) periodic active fio in production, (b) passive-only, (c) active-on-idle + passive-in-production. Decision: (c). Consequences: benchmarking is bounded to empty/pre-user servers; provisioning/QC decisions read passive metrics; this ADR prevents the active-production-fio idea (this issue) from recurring.
- **RECOMMEND_NEW_ISSUE** — Draft title: *"Post-install idle storage baseline: auto-run storageBenchmark.php (--require-idle) at install completion, guarded against existing users."* Scope: the safe Slice 1 above, decoupled from the dangerous/merge parts of #51.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Investigation Report (2nd pass — verification + corrections)

**Issue:** #51 — Variety of automatic Benchmarks after installation, and potentially on-going QC
**Investigated:** 2026-06-04
**Status:** PARTIAL

> This is a **second-pass verification** of the thorough investigation posted on 2026-06-03. Rather than
> re-derive it, this pass independently verified that report's load-bearing claims against current source,
> corrects two points that had drifted, and drives toward a decision. A third full investigation would be
> redundant — the question is now a product decision, not a research gap.

### Summary
The prior report's central claims **hold up against source**: the storage-benchmark engine
(`scripts/lib/storageBenchmark.php`) already implements the exact fio profile this issue asks for, with a
fail-closed idle gate; and passive ongoing-QC already runs on a cron cadence. Two corrections: (1) the GUI/
result-graph residue should merge into **#501**, not #45 — #45 now carries the `should-not-implement` label and
#501 is its live successor ("…closes #45"); (2) the reason to refuse ask #5 is a general multi-tenant I/O-
saturation hazard — aggressive `fio --direct=1` at high iodepth on a *populated* host can starve shared storage,
spike host load, and trigger kernel soft-lockups that affect co-located workloads — so active benchmarking must
be confined to empty/pre-user servers. Recommendation remains **SPLIT_FIRST**, with the dangerous active-
production ask refused *as specified* (operator may revisit a guardrailed variant), and the safe post-install
slice spun out.

### What The Issue Asks For
1. Post-install benchmarks (`yabs.sh` + a network speed test), logged.
2. A cron loop running every 60 min "until there are users in the server".
3. Long-term, user-viewable result storage (datetime, duration).
4. fio I/O stress: 95% read / 5% write, random 512–1024 KiB, iodepth 32–100, **randomized settings**, **mixed
   sequential-load injection** (pigz/tar, stream to `/dev/null`, dd), **multiple concurrent fio instances**, with
   iostat captured during the run.
5. **Active fio in production every few weeks**, and **auto-disable new users** when performance is too low (a
   provisioning gate).
6. GUI indicator while a perf test runs; ~2-min weekly / ~10-min monthly cadence.

### Implementation Status
Verified by reading source this run:
- **`scripts/lib/storageBenchmark.php`** (+ `scripts/lib/storageBenchmark/report.php`, util wrapper) implements
  ask #4's single-job profile **verbatim**: `randmix-large-95r5w` = `rw=randrw rwmixread=95
  bssplit=4k/2:64k/3:128k/5:256k/10:512k/20:768k/25:1024k/35 iodepth=32 numjobs=4 direct=1`, alongside
  randread-1M, randread-4k, randwrite-4k, seqread-1M.
- **Idle gate is real and fail-closed:** `--require-idle` aborts ("Busy system … aborting", non-zero exit) when
  ioping/iostat exceed thresholds (defaults: latency 100 ms, util 85%). Per-device tests add cross-device
  outlier warnings. Input/path hardening is present.
- **JSONL log** at `/var/log/pmss/benchmark-storage.jsonl` (run id, timestamp, params, read/write MB/s, IOPS,
  p95) — programmatically accessible; `--show-last` renders the last run.
- **Passive ongoing QC already runs via `etc/seedbox/config/root.cron`:** resource log (5 min), disk iostat
  (5 min), cpu stat (15 min), system-stats with ioping + PSI (5 min), process snapshot (30 min), SMART snapshot
  (2×/day), and a quarterly mdadm `checkarray`.

### What's Missing
- No install-time trigger and no "until there are users" pre-user loop (`scripts/lib/user/selection.php` is a
  ready reuse point for the no-users check).
- `yabs.sh` and a network-throughput test are absent.
- No GUI integration and no rollup/graph for the JSONL (that residue belongs to the #501/#45 metrics line).
- The benchmark runs a **fixed job list sequentially** (confirmed: a single `foreach` over the job list) — no
  concurrent fio instances, no randomized settings/durations/datasets, no mixed sequential-load injection.
- No active-production-fio QC and no auto-disable-new-users provisioning gate.

### Mission Alignment
- **Reliability over heroics / business longevity:** the existing tool + passive cron deliver QC without risking
  live service. Ask #5 (active fio in production) directly endangers reliability (see Related Incidents).
- **Investigate then act:** the right provisioning signal is *observed contended load* (passive PSI / service
  time / iostat), not a synthetic idle-time peak. Building the active mechanism is acting before the passive
  evidence is in.
- **Honesty / no hallucination:** ask #5 was operator-authored; the honest move is to surface *why it is unsafe
  as written* and let the operator decide — not to silently drop it.

### Devil's Advocate
*Why not build #51 as written?* Most of it already exists; the dangerous part recreates a known I/O-saturation
hazard on populated multi-tenant hosts; and `yabs.sh` is a third-party remote-script-piped-to-shell installer
that conflicts with the shell-execution-safety
posture (ADR-0004/0005) while duplicating the disk benchmark already shipped. *Rebuttal for the slice worth
building:* a one-time **post-install idle baseline** is genuinely useful (acceptance test + capacity baseline +
per-device outlier catch) and is safe precisely because there are no users yet.

### Anti-Pattern Hunt
- REGEX_FOR_SEMANTICS: PASS — metrics are numeric; no semantic classification.
- PREMATURE_OPTIMIZATION: PASS — recommendation rejects the speculative concurrent/randomized fio orchestration
  and the production-QC/provisioning machinery (no demonstrated need in ~2 years); only a small, justified slice
  is endorsed.
- FALSE_COMPLETION_CLAIM: PASS — status PARTIAL; every "done" claim cites a file read this run; gaps enumerated.
- OPERATOR_HANDOFF_ANTIPATTERN: PASS — primary recommendation is SPLIT_FIRST with concrete per-slice verdicts and
  a spun-out actionable slice, not a deferral.
- FABRICATED_AUTHORITY_OR_PATH: PASS — every cited path (`storageBenchmark.php`, `root.cron`, `docs/adr/`,
  `user/selection.php`) verified this run; issue states (#45/#501) verified via API.
- FAIL_OPEN_SAFETY_BEHAVIOR: PASS — the endorsed slice enforces `--require-idle` (verified fail-closed: aborts
  on busy) and explicitly refuses the run-while-busy path.
- DRY_RUN_THEATER: PASS — investigation read real source + real incident records.
- INVESTIGATION_LOOP_THEATER: PASS (actively guarded) — this pass does **not** re-derive the prior report; it
  verifies it, corrects two drifted points, and pushes to a decision. It explicitly flags that a third full pass
  would be theater.

### Related Incidents & Data
- **Known I/O-saturation hazard (general):** running aggressive `fio --direct=1` at high iodepth on a *populated*
  multi-tenant host can saturate shared storage, spike host load, and in the worst case trigger kernel
  soft-lockups that degrade co-located workloads. The benchmark's default jobs use exactly this shape
  (`direct=1 iodepth=32 numjobs=4`) — safe on an empty/pre-user server, hazardous on a populated one. This is the
  central reason to refuse ask #5 as specified and the reason the tool ships a fail-closed `--require-idle` gate.
  (Specific operational incident records that motivated this are tracked privately, not on this public issue.)
- **Frequency / pressure:** #51 has sat open ~2 years with no operational pressure for the active-fio part, while
  the passive-metrics line (#501) is actively progressing — a low-frequency signal for the speculative parts.

### Cross-Issue Analysis
- **#501** (9 missing per-user/VM metrics + iostat-history persistence; "closes #45") — **OPEN, live successor.**
  This is the correct home for both the provisioning-headroom signal and the GUI/result-graph residue of #51.
  **MERGE / DEPENDS.**
- **#45** (local long-term I/O performance stats) — OPEN **but labeled `should-not-implement`**. *Correction to
  the prior report:* do **not** route #51's residue here; route it to #501.
- **#121 / #437** (cgroup IO/burst throttle + OOM logging) — OPEN; the in-guest I/O containment that makes any
  intentional load generation safer. Landing order: containment before anything that deliberately generates load.
- No file-level edit conflict: no other open issue touches `storageBenchmark.php`.

### Dependencies & Blockers
- Slice 1 (post-install idle baseline): no blockers — the tool exists; needs only an install/cron hook + a
  no-users guard.
- GUI/result-graph: depends on #501's rollup/stats work (in progress — not yet a shipped fact).
- Any future active-benchmarking variant: depends on verified cgroup throttle coverage (#121/#437/#273) and an
  explicit operator decision.

### Reuse, Refactor & Test Strategy
- **Existing reuse points:** `scripts/lib/storageBenchmark.php` (engine + fail-closed idle gate + JSONL),
  `scripts/lib/user/selection.php` (no-users detection), `etc/seedbox/config/root.cron` (scheduling slot), the
  passive cron collectors above.
- **Refactor opportunities:** add a thin `--require-idle`-enforced invocation hook rather than new benchmark
  logic (DRY); fold any result display into the existing stats/metrics surface (collection vs presentation are
  already separated between the engine and `report.php`).
- **Test categories proposed:** hermetic unit for the no-users gate decision; reuse the existing benchmark test
  pattern for any new option; **regression** test that the install hook refuses to run when users exist / the
  system is busy (guards against re-triggering the March-2026 failure mode); boundary test for "exactly one user
  present"; failure-injection test that a busy system aborts with non-zero exit.
- **Minimum-code-for-correctness:** an install-completion call to the existing tool with `--require-idle`, plus a
  user-count guard. No new abstraction, no configurability without a second caller.
- **Mission/business success signal:** a captured per-server idle baseline in the JSONL at provisioning time
  (acceptance evidence before a server takes customers); business signal: catches dud disks/arrays **before** a
  paying customer lands on them (retention + avoided refund/migration cost).

### Recommendation
**SPLIT_FIRST** (confirmed from the prior pass, refined):
- **Slice 1 — IMPLEMENT (small, safe):** post-install idle baseline — invoke the existing
  `storageBenchmark.php` with `--require-idle` at install completion; optional pre-user repeat loop guarded by a
  no-users check. One-time peak-capacity baseline + per-device outlier check on the empty server.
- **Slice 2 — SHOULD-NOT-IMPLEMENT *as specified*:** active fio in production + auto-disable-new-users gate.
  It recreates a known I/O-saturation hazard on populated multi-tenant hosts. The real provisioning-headroom need
  is met by the passive-metrics
  line (#501) + a cgroup-based saturation gate. *This was operator-authored — so it is not unilaterally killed:*
  any future variant is **gated** on (a) operator re-approval, (b) verified cgroup throttle coverage + dedicated
  per-VM arrays, (c) proof passive metrics are insufficient.
- **Slice 3 — drop the mechanism, keep the need:** `yabs.sh` (remote-script-piped-to-shell, conflicts with
  shell-safety ADRs; disk part duplicates the existing tool). If a CPU/RAM/network post-install baseline is
  wanted, implement a small local equivalent — don't pipe a third-party script.
- **Slice 4 — MERGE into #501:** result GUI / stats-page graph for the benchmark JSONL.
- **Ask #4 orchestration** (concurrent multi-instance, randomized settings/durations, mixed-load injection) =
  gold-plating; no demonstrated need in ~2 years — don't build absent a concrete driver.

Reasoning: most of this 2024 issue is already shipped; the remainder is one safe small slice plus parts that
should be refused-as-specified or routed to their proper home (#501). SPLIT_FIRST keeps operator visibility on
the decomposition rather than auto-closing a partly-valid, operator-authored request.

### Auto-Close Decision
AUTO_CLOSED_BLOCKED: trigger-not-met; reason=recommendation-is-IMPLEMENT/NEEDS_RESEARCH/NEEDS_OPERATOR_INPUT/SPLIT_FIRST

### Additive Recommendation Flags
- **RECOMMEND_ADR** — Draft title: *"Performance QC and provisioning gating are passive-metric-driven; active fio
  benchmarks run only on idle/pre-user servers, never in production with live users."* Context: PMSS already
  collects passive I/O telemetry (PSI/ioping/iostat) on cron and ships a non-destructive `storageBenchmark.php`;
  aggressive fio on a populated multi-tenant host is a known I/O-saturation hazard. Options: (a)
  periodic active fio in production, (b) passive-only, (c) active-on-idle + passive-in-production. Decision: (c).
  Consequence: this ADR prevents the active-production-fio idea from recurring as a fresh issue.
- **RECOMMEND_NEW_ISSUE** — Draft title: *"Post-install idle storage baseline: auto-run storageBenchmark.php
  (--require-idle) at install completion, guarded against existing users."* Scope: the safe Slice 1, decoupled
  from the dangerous/merge parts of #51. (After this spin-out, #51 can be closed as decomposed.)

### Deep Research Gate
SKIPPED: prior-investigation-count — #51 has now been investigated 2× (2026-06-03 + this pass); the engine,
cron pipeline, the fio-saturation incident records, and the issue states were all read directly this run.
Internal evidence is sufficient; no novel domain or industry comparison is required.

### Architectural Minimization
**Step A — dumbest requirement:** "run fio in production every few weeks and auto-disable new users on low
performance." Author = the operator, **in 2024** — before passive PSI/ioping/iostat collection existed and before
the 2026 fio-saturation incident. The underlying constraint (provisioning must respect I/O headroom) is real and
current; the *active-fio mechanism* is an inherited 2024 assumption that newer passive instrumentation made both
unnecessary and unsafe.

**Step B — alternatives:**
- **DELETE:** (1) Delete ask #5's active-production-fio — the headroom signal is provided by the passive cron
  metrics already collected (preferred path for the production-QC portion). (2) Delete `yabs.sh`/network-pipe —
  disk part duplicates `storageBenchmark.php`; a remote-script-piped-to-shell installer conflicts with
  shell-execution-safety.
- **MERGE:** (1) Merge #51's result-GUI into #501's stats/rollup work (shared consumer of perf data). (2) Merge
  the performance-provisioning gate into the cgroup/metrics line (#121/#437/#501).
- **REFACTOR:** (1) Refactor the install path to call the existing `storageBenchmark.php` once at completion
  (`--require-idle`). (2) Reuse `scripts/lib/user/selection.php` as a no-users guard for an optional pre-user loop.
- **NULL:** If nothing is done — no automatic fresh-server I/O baseline is captured; operators run the benchmark
  manually during investigations; passive QC + capacity planning continue. Concrete consequence: per-server
  acceptance/regression comparison stays manual. Modest impact.

**Step C — minimal design proposal:** Add an install-completion (and optional pre-user-loop) invocation of the
existing `storageBenchmark.php` with `--require-idle` enforced and a user-count guard from
`scripts/lib/user/selection.php`; results already land in the JSONL. Files: a small hook in the install/update
completion path + one cron-loop guard (~30–60 LOC). Stripped vs the issue: active production fio, the
auto-disable gate, yabs/network-pipe, the concurrent/randomized orchestration, and a bespoke GUI — each stripped
because it is dangerous, redundant, speculative, or belongs to #501.

### ADR Adherence
NO_RELEVANT_ADRS govern storage benchmarking directly: listed `docs/adr/` (0001, 0003–0024); none covers
performance benchmarking/QC. The recommendation is consistent with the observability-oriented decisions and does
not conflict with the cgroup ADRs (0003 v2 adoption, 0019 v1 pin) or 0023 (cron-service TasksMax guard) — the
endorsed slice adds one idle-gated invocation, not a new resource-control mechanism. It *aligns with* the
shell-execution-safety posture (0004/0005) by rejecting the remote `yabs.sh` pipe.

### Final AQFM Sweep
Question asked (verbatim):
> "Issue #51 (MagnaCapax/PMSS): Variety of automatic Benchmarks after installation and ongoing QC. This is a
> 2nd-pass verification of a 2026-06-03 investigation. Draft conclusion: storageBenchmark.php already implements
> the 95r5w fio profile and a fail-closed --require-idle gate; passive QC runs via root.cron; aggressive
> fio --direct=1 on a populated multi-tenant host is a known I/O-saturation hazard (so confine active
> benchmarking to empty/pre-user servers); issue #45 is labeled should-not-implement so the GUI/metrics
> merge target is #501 not #45. Draft recommendation: SPLIT_FIRST — implement only a post-install idle baseline,
> SHOULD-NOT-IMPLEMENT active production fio + auto-disable provisioning gate, drop yabs, merge GUI into #501.
> What contradicting evidence, missed operator decision, or relevant research could revise this draft?"

Verdict: **REVISED** — three refinements adopted: (1) ask #5 was operator-authored → don't flatly kill it; refuse
*as specified* and **gate** any future guardrailed variant on operator re-approval + verified cgroup coverage +
passive-metrics-insufficiency; (2) `yabs.sh` covers CPU/RAM/network too → drop the *remote-pipe mechanism*, not
the baseline *need*; (3) the passive-metrics path (#501) is **in progress, not shipped** → it is a dependency, not
a current fact. Unverified specific figures were not adopted into this public report.

### Adversarial Self-Challenge
**Strongest finding 1:** `storageBenchmark.php` implements ask #4's fio profile and a fail-closed idle gate.
- Disproving search: re-read the job array + `fioRun` + the `--require-idle` branch and the run loop. → The job
  profile matches verbatim and the idle gate aborts non-zero on busy; but the run loop is a single sequential
  `foreach` — **no concurrency/randomization/mixed-load**, so ask #4's *orchestration* is genuinely unbuilt.
- Outcome: **held, with the orchestration gap explicitly carved out** (reflected in What's Missing).

**Strongest finding 2:** running aggressive fio in production is hazardous (I/O saturation on populated hosts).
- Disproving search 1: internal incident records + their cause analysis. → Confirmed the hazard is real and
  documented; the precise causal chain is more nuanced than "fio alone", but the saturation risk on a populated
  host stands.
- Disproving search 2 (broader): searched for any record advocating periodic active production benchmarking as
  desirable. → Found only passive-fix recommendations for the provisioning need; none advocate active fio.
- Outcome: **held** — hazard real; the alternative is passive.

**Least-evidenced finding** (lowest citation count): "the GUI/metrics residue should merge into #501, not #45."
- Disproving search: queried both issues' live state via API. → #45 is OPEN **with the `should-not-implement`
  label**; #501 is OPEN and titled "…closes #45". So #45 is not a valid merge target; #501 is.
- Outcome: **held** — this is the correction to the prior report.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Investigation Report (3rd pass — verification + loop-break, NOT a re-derivation)

**Issue:** #51 — Variety of automatic Benchmarks after installation, and potentially on-going QC
**Investigated:** 2026-06-04
**Status:** PARTIAL

> This issue already carries two thorough investigations (2026-06-03 and 2026-06-04 above), both
> converging on **SPLIT_FIRST**. The 2nd pass explicitly stated it was driving to closure "so a 3rd
> pass never happens." The automated pipeline re-selected #51 anyway. Re-deriving the full report a
> third time is the `INVESTIGATION_LOOP_THEATER` anti-pattern. So this pass does only two honest things:
> (1) **independently re-verify the load-bearing claims against current source** (two agreeing reports
> can still both be wrong), and (2) **name the loop and drive the single decision that ends it.**

### Independent re-verification (this pass, against current source)
All confirmed — no drift since the prior passes:
- **Core fio engine exists, ask-#4 profile verbatim:** `scripts/lib/storageBenchmark.php:89` job
  `randmix-large-95r5w` = `rw=randrw rwmixread=95 bssplit=4k/2:64k/3:128k/5:256k/10:512k/20:768k/25:1024k/35
  iodepth=32 numjobs=4 direct=1`. Fail-closed idle gate at `:59` (aborts non-zero on busy; defaults latency
  100 ms / util 85 % at `:46`). JSONL results log; `--show-last` renderer + tests present.
- **Sequential only:** `:102` / `:145` are plain `foreach` loops — confirmed **no** concurrency, randomized
  settings/durations, or mixed sequential-load injection (ask #4's orchestration layer is genuinely unbuilt).
- **Manual-invocation only:** no caller of the benchmark exists in `install.sh`, cron, or the update path —
  only the util wrapper and tests reference it. So the "post-install trigger" gap holds.
- **Passive ongoing QC already runs** via `etc/seedbox/config/root.cron`: resource/iostat (5 min), cpu (15 min),
  system-stats with ioping+PSI (5 min), process snapshot (30 min), SMART snapshot (2×/day), quarterly checkarray.
- **Cross-issue states (gh API, today):** #45 OPEN but labeled `should-not-implement`; #501 OPEN, live successor
  ("…closes #45"); #378 OPEN. → GUI/metrics residue routes to **#501, not #45** (the 2nd-pass correction holds).

**Verdict on the prior conclusion: confirmed correct against source.** Nothing in the technical analysis needs
revision. The decomposition stands: core = shipped; active-prod-fio + auto-disable gate = refuse *as specified*
(operator-authored — revisit only as a guardrailed, cgroup-gated variant if ever wanted); `yabs.sh`/remote-pipe =
drop the mechanism, keep any CPU/RAM/net baseline need; GUI/result-graph = merge into #501; concurrent/randomized
fio orchestration = gold-plating with no demonstrated driver in ~2 years.

### The actual finding of this pass — the loop is the problem, not the analysis
#51 has now been investigated three times with identical conclusions. The reason it keeps returning is structural,
not analytical: a `SPLIT_FIRST` recommendation has no auto-action path (the pipeline cannot create the spin-out
issue, and auto-close intentionally excludes `SPLIT_FIRST`), and the candidate pool does not yet exclude
resolved-but-open issues — which is exactly **#378**. Until either the operator decomposes #51 or #378 lands, a
4th pass is the predictable next event. Posting a 4th identical SPLIT_FIRST would feed that.

### Anti-Pattern Hunt
- INVESTIGATION_LOOP_THEATER: **the active concern** — handled by refusing to re-derive; this pass verifies +
  escalates the loop instead. The genre is documented (cf. #497's 5th pass, #418's repeated passes).
- FALSE_COMPLETION_CLAIM: PASS — status PARTIAL; each "done" claim re-cited to a file:line read this pass.
- PREMATURE_OPTIMIZATION: PASS — the speculative orchestration + production-QC machinery is explicitly *not*
  endorsed; only the small post-install idle baseline survives the cut.
- OPERATOR_HANDOFF_ANTIPATTERN: PASS (with care) — this is not "the operator should decide" laziness; the
  technical work is done and verified 3×, and the *only* remaining steps (create spin-out, close, the active-fio
  product call) are authority-gated and outside a read-only pass.
- FABRICATED_AUTHORITY_OR_PATH: PASS — every path/line cited was read this pass.
- FAIL_OPEN_SAFETY_BEHAVIOR: PASS — `--require-idle` verified fail-closed (non-zero exit on busy).
- REGEX_FOR_SEMANTICS / DRY_RUN_THEATER: PASS — n/a (numeric metrics; real source reads).

### Recommendation
**NEEDS_OPERATOR_INPUT** — one decision ends a 2-year-old, 3×-investigated loop:

> **Close #51 as decomposed.** Its content is fully decided and verified: the core benchmark is shipped; the
> dangerous active-production-fio + auto-disable-new-users ask is refused *as specified*; `yabs`/remote-pipe is
> dropped; GUI/metrics belongs to #501. The single buildable remainder — a **post-install idle storage baseline**
> (invoke the existing `storageBenchmark.php --require-idle` at install completion, guarded by a no-users check) —
> wants its own small issue so it isn't orphaned by the close.

Two operator actions resolve it permanently: (a) bless the spin-out + close #51; (b) #378 (candidate-pool
exclusion) so resolved-but-open issues stop being re-selected. The active-fio slice is the only genuine *product*
decision and is gated, per the 2nd pass, on operator re-approval + verified cgroup coverage + proof the passive
metrics line (#501) is insufficient.

### Auto-Close Decision
AUTO_CLOSED_BLOCKED: trigger-not-met; reason=recommendation-is-IMPLEMENT/NEEDS_RESEARCH/NEEDS_OPERATOR_INPUT/SPLIT_FIRST
(SPLIT_FIRST/NEEDS_OPERATOR_INPUT are not auto-close categories, and the post-install-baseline slice has no
tracking issue yet — auto-closing now would orphan the one buildable part. Closure is the operator's call above.)

### Additive flags
- **RECOMMEND_NEW_ISSUE** — *"Post-install idle storage baseline: auto-run `storageBenchmark.php --require-idle`
  at install completion, guarded against existing users."* (the safe slice; after spin-out, #51 closes clean.)
- **RECOMMEND_ADR** — record that performance-QC/provisioning is passive-metric-driven and active fio runs only
  on idle/pre-user servers, never on populated multi-tenant hosts — so this idea cannot recur as a fresh issue.

*Depth note:* the exhaustive sections (full Architectural Minimization, ADR list, AQFM sweep) are in the two
prior comments and were re-confirmed, not re-typed, here — duplicating them is the theater this pass exists to
avoid.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Correction to the comment above — recommendation is SPLIT_FIRST, not NEEDS_OPERATOR_INPUT

I have to correct my own previous comment. I labelled the recommendation `NEEDS_OPERATOR_INPUT`; that was
wrong, and I'm flagging it rather than leaving it to stand.

**What I got wrong and why.** I let the goal of "stop this issue from being re-investigated" bend the verdict
label. That is backwards: every ask in this issue already has a clear, determinable disposition, so deferring to
the operator is the operator-handoff antipattern — the exact thing the two prior passes explicitly avoided
("primary recommendation is SPLIT_FIRST with concrete per-slice verdicts, not a deferral to the operator").

**The accurate verdict — SPLIT_FIRST** (unchanged from, and consistent with, both prior passes; the underlying
source verification in my comment above still holds):
- **Slice 1 — IMPLEMENT (small, safe):** post-install idle baseline — call the existing `storageBenchmark.php`
  with `--require-idle` at install completion, guarded by a no-users check. Spin out as its own small issue so
  it isn't orphaned when #51 closes.
- **Slice 2 — SHOULD-NOT-IMPLEMENT (as specified):** active fio in production + auto-disable-new-users gate. The
  real provisioning-headroom need is served by passive metrics (#501), not by synthetic load on live
  multi-tenant hosts. A guardrailed future variant stays gated on operator re-approval + verified cgroup
  coverage (#121/#437) + proof passive metrics are insufficient.
- **Slice 3 — drop the mechanism:** `yabs.sh`/remote-pipe (conflicts with the shell-execution-safety posture,
  ADR-0004/0005; disk part duplicates the existing tool). Keep any CPU/RAM/net baseline *need* if wanted.
- **Slice 4 — MERGE into #501:** result GUI / stats-graph for the benchmark JSONL.
- Concurrent/randomized fio orchestration (part of ask #4) = gold-plating; no driver in ~2 years.

**The loop is a separate matter, handled at its root, not by relabeling this verdict.** This issue has now been
investigated three times with the same conclusion because a `SPLIT_FIRST` verdict has no auto-action path and the
candidate pool does not yet exclude resolved-but-open issues — that is **#378**, a pipeline concern, not a reason
to distort #51's verdict. Resolution: land #378, and decompose #51 (spin out Slice 1, then close as decomposed).

### ADR Adherence (refinement)
EXTENDS: ADR-0007 (install bootstrap contract) and ADR-0006 (update-step2 user-loop + observability) — the Slice-1
install-completion hook builds on those existing install/observability decisions rather than introducing a new
mechanism. No ADR governs storage benchmarking directly; none is in conflict.

### Auto-Close Decision
AUTO_CLOSED_BLOCKED: trigger-not-met; reason=recommendation-is-IMPLEMENT/NEEDS_RESEARCH/NEEDS_OPERATOR_INPUT/SPLIT_FIRST
(SPLIT_FIRST is not an auto-close category, and the Slice-1 spin-out issue does not exist yet — auto-closing now
would orphan the one buildable part.)

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Investigation Report (5th pass — source verification, not a re-derivation)

**Issue:** #51 — Variety of automatic Benchmarks after installation, and potentially on-going QC
**Investigated:** 2026-06-04
**Status:** PARTIAL · **Recommendation:** SPLIT_FIRST (unchanged)

This issue already carries four investigation comments converging on **SPLIT_FIRST**. Rather than re-derive that analysis (it is complete in the comments above), this pass does the one useful thing left: **independently re-verify the load-bearing claims against current source** — four agreeing reports can still all be wrong.

### Re-verified against current source today
- **Core fio engine is shipped, matching ask #4 verbatim:** `scripts/lib/storageBenchmark.php` job `randmix-large-95r5w` = `rw=randrw rwmixread=95 bssplit=4k/2:64k/3:128k/5:256k/10:512k/20:768k/25:1024k/35 iodepth=32 numjobs=4 direct=1`, alongside randread (1M / 4k), randwrite (4k), seqread (1M). Fail-closed idle gate (`--require-idle`, defaults 100 ms latency / 85 % util). Results land in a JSONL log; `--show-last` renders them. **The engine file was modified today** — it is under active development.
- **Post-install trigger still absent:** no caller of the benchmark exists in `install.sh`, the update path, or cron — only the util wrapper and tests reference it. The one genuinely buildable slice (a post-install idle baseline) remains unbuilt.
- **Passive ongoing-QC already runs** via `etc/seedbox/config/root.cron`: resource/iostat/cpu logging, system-stats with ioping + PSI, process and SMART snapshots, and a periodic mdadm `checkarray`.
- **Cross-issue states:** #45 OPEN but labelled `should-not-implement`; #501 OPEN, live successor ("closes #45"); #378 OPEN. → GUI/metrics residue belongs to **#501**.

**Verdict against source: confirmed. Nothing has drifted since the prior passes.**

### Disposition (unchanged, SPLIT_FIRST)
- **IMPLEMENT (small, safe):** post-install idle baseline — call the existing `storageBenchmark.php --require-idle` at install completion, guarded by a no-users check. Spin out as its own small issue.
- **SHOULD-NOT-IMPLEMENT (as specified):** active fio in production + auto-disable-new-users gate. Aggressive fio at high iodepth on a *populated* multi-tenant host is a known I/O-saturation hazard; the provisioning-headroom need is better met by the passive-metrics line (#501) + cgroup throttling (#121 / #437), not by synthetic load on live customers. A future guardrailed variant stays gated on operator re-approval + verified containment.
- **Drop the mechanism:** `yabs.sh` / remote-pipe (conflicts with the shell-execution-safety posture; its disk part duplicates the existing tool). Keep any CPU/RAM/network baseline *need* if wanted, implemented locally.
- **MERGE into #501:** result GUI / stats-graph for the benchmark JSONL.
- Concurrent/randomized multi-instance fio orchestration (part of ask #4) = gold-plating; no demonstrated driver in ~2 years.

### Anti-Pattern Hunt
- INVESTIGATION_LOOP_THEATER: actively guarded — this pass verifies against current source and does **not** re-type the prior analysis. #51 has now been examined five times; the recurrence is structural, not analytical — a `SPLIT_FIRST` verdict has no auto-action path, and resolved-but-open issues aren't yet excluded from the candidate pool (that is #378).
- FABRICATED_AUTHORITY_OR_PATH: PASS — every path cited was read this pass.
- FAIL_OPEN_SAFETY_BEHAVIOR: PASS — `--require-idle` verified fail-closed (non-zero exit on a busy system).
- Remaining classes (regex-for-semantics, premature-optimization, false-completion, operator-handoff, dry-run): PASS — established in the prior passes; numeric metrics, real source reads.

### Recommendation
**SPLIT_FIRST.** The clean way to end the recurrence: spin out the safe post-install-baseline slice as its own issue, then close #51 as decomposed; land #378 so decided-but-open issues stop being re-selected.

### Auto-Close Decision
AUTO_CLOSED_BLOCKED: trigger-not-met; reason=recommendation-is-IMPLEMENT/NEEDS_RESEARCH/NEEDS_OPERATOR_INPUT/SPLIT_FIRST
(SPLIT_FIRST is not an auto-close category, and the post-install-baseline slice has no tracking issue yet — auto-closing now would orphan the one buildable part.)

### Additive Flags
- **RECOMMEND_NEW_ISSUE** — *"Post-install idle storage baseline: auto-run `storageBenchmark.php --require-idle` at install completion, guarded against existing users."* The safe slice; after spin-out, #51 closes clean.
- **RECOMMEND_ADR** — record that performance-QC / provisioning is passive-metric-driven and active fio runs only on idle / pre-user servers, never on populated multi-tenant hosts — so this idea cannot recur as a fresh issue.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Verification pass — re-checked against current `main`; verdict unchanged (PARTIAL / SPLIT_FIRST)

Brief re-verification, **not** a re-derivation — the full analysis and per-slice rationale are in the comments above. I re-read the current source today; the picture holds, with one fresh observation.

**Re-verified against current source:**
- **Core fio engine ships.** `scripts/lib/storageBenchmark.php` (`storageBenchmarkFileJobs()`) carries the requested seedbox-style profile verbatim — `randrw`, 95% read / 5% write, `bssplit 4k…1024k`, `iodepth=32`, `numjobs=4`, `direct=1` — plus sequential and small-random jobs and device-level jobs.
- **Idle gate present and fail-closed.** `--require-idle` (defaults `--idle-latency-ms 100`, `--idle-util 85`) aborts the run on a busy system.
- **Still no automatic trigger.** No caller in `update.php`, the install path, or `root.cron` — the benchmark is manual-only today, so "run automatically after installation" remains the one slice not yet wired.
- **Passive on-going QC already exists** via the periodic collectors in `root.cron` (resource / iostat / PSI / ioping / SMART snapshots), covering much of the "on-going QC" intent without a heavy synthetic load.
- The benchmark file is under active refactoring as of today, so the engine is live and maintained.

**Disposition (unchanged):**
- **Ship:** an idle-gated, no-users post-install baseline — small, since the idle-gate primitive already exists; only the trigger wiring is missing.
- **Keep gated / defer:** running heavy `fio` in production against servers that already have users, with automatic "disable new users if performance is low." Concurrent heavy synthetic I/O on a host with live tenants can saturate shared I/O and degrade other users — this belongs behind enforced per-user I/O limits and explicit sign-off, not an unconditional schedule.
- **Drop the mechanism:** piping remote scripts (`yabs.sh` / external speed-test) conflicts with shell-execution safety; any CPU/RAM/network baseline can be produced in-tree.
- **Merge:** user-visible result graphs and long-term result storage → tracked by #501.
- **Gold-plating, no demand:** randomized concurrent multi-instance fio orchestration.

**Recommendation:** the umbrella is largely resolved-or-intentionally-deferred per slice, and the core engine has already landed. Suggest **decomposing #51** — spin the post-install idle-baseline slice into its own small issue and close this umbrella — so the remaining buildable work (just that trigger) is tracked discretely.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review

**Issue:** #51 — Variety of automatic Benchmarks after installation, and potentially on-going QC
**Reviewed:** 2026-06-12
**Verdict:** SHOULD-NOT-IMPLEMENT *(umbrella as written)* — decomposed; the one buildable slice is now spun out as **#656**

This is the pre-implementation gate. The prior analysis on this thread converged on a per-slice split; this review independently re-verified every load-bearing claim against current source, adds two refinements, and takes the terminal action (spin-out + label) so the implementation decision is unambiguous. **No part of #51 as written should be implemented** — most of it is already shipped, the headline ask is unsafe, and the single buildable remainder now lives in its own issue.

### Adversarial Findings

**Big Questions (Munger inversion + Musk first-principles):**

1. **Worst case if implemented as written.** Ask #5 ("run fio in production every few weeks; auto-disable new users when performance is low") schedules the heaviest job — `randmix-large-95r5w` (`iodepth=32 numjobs=4 direct=1`, `storageBenchmark.php:98`) — against hosts that already carry tenants → heavy synthetic I/O competes for the host's CPU/controller/storage bandwidth → load runaway / kernel soft-lockup → co-located tenant services freeze → multi-hour outage detected only when customers complain → recovery needs a manual host power-cycle and per-tenant verification. The tool ships a **fail-closed `--require-idle` gate** (`:57`, `return 2` on busy) precisely because this shape is dangerous on a populated host.
2. **Inversion (guarantee maximum damage).** You would schedule that exact job unconditionally on tenant-loaded hosts with no idle gate, then wire it to auto-disable provisioning so the host also stops earning while degraded. Ask #5 differs from this worst case only by the (unbuilt) cadence and the (deliberately bypassed) idle gate. The distance between "ask #5" and "the incident generator" is one missing `--require-idle`.
3. **Regret in 6 months.** A synthetic-load QC path that fires on live tenants becomes the thing you must remember to disable during every incident; and an auto-disable gate keyed on a synthetic *idle-time peak* mis-gates provisioning because the idle ceiling is not the contended-load headroom you actually need to decide "can this host take one more user."
4. **Removal inversion.** If the post-install baseline already existed, would we fight to keep it? **Yes — if** it surfaces a slow/dud disk at install. **No — if** it only appended JSONL nobody reads. That fork is the load-bearing design constraint for #656 (see refinement A).
5. **Dumbest requirement (Musk Step 1).** "Run fio in production every few weeks and auto-disable new users on low performance" — it endangers the very tenants it means to protect, and it measures an idle-time ceiling rather than real contended headroom. It is a 2024 requirement written before passive load telemetry existed.
6. **Deletion opportunity (Musk Step 2 / "best part is no part").** Instead of building the umbrella, delete most of it from the queue: the fio engine is shipped; the "ongoing QC" intent is already met by the passive cron collectors (`root.cron`: `resourceLog`/`diskIostat`/`cpuStat`/`systemStatsLog` ioping+PSI/`storageHealthSnapshot`). Only the post-install trigger remains — and that is now #656, not this umbrella.

**Refinement A — the silent-WARN trap (new).** The slow-disk signal at install time is `storageBenchmarkPrintPeerWarnings()` (`storageBenchmark.php:126`), which prints `WARN:` lines to **stdout during the run**. Those lines are **not** written to the JSONL and **not** reconstructed by the `--show-last` renderer (`storageBenchmark/report.php`). An install hook that only relies on the JSONL append would silently lose exactly the signal that justifies running at install. #656 must capture stdout to an operator-visible log, or the feature collects data for its own sake.

**Refinement B — delete the 60-minute loop (new).** Ask #2 wants the benchmark re-run every 60 min "until there are users." A **single** post-install run is sufficient: idle-server disk performance is stable, and drive degradation is already covered by the 2×/day SMART snapshot cron. A repeat loop adds a provisioning race (a user can be created mid-run) and a transient `/home` fill risk (the file test allocates up to 80% of free space) for ~zero added signal. #656 specifies a single run.

### Caller Analysis
#51 **adds** functionality; it removes nothing, so the removal/caller-enumeration hard gate is N/A. The enumeration that matters is the reverse — what callers each ask would *add*:
- **Ask #5 (active prod fio):** adds a scheduled caller of the heaviest fio jobs against **populated** hosts. This is the dangerous new caller. Refuse as specified.
- **#656 (post-install baseline):** adds **one** guarded caller in `update-step2` — `--require-idle` enforced + no-users guard. Safe by construction (runs only on empty/idle hosts).
- No existing caller of `storageBenchmark` exists outside its util wrapper, report module, tests, and a lint script (verified by `grep` across `install.sh`, `update.php`, `util/update-step2.php`, `etc/seedbox/config/root.cron`).

### Investigation Results
Independently re-verified against current `main` (line numbers current post-decomposition-refactor):
- **Ask #4 fio profile is shipped verbatim:** `storageBenchmarkFileJobs()` `storageBenchmark.php:98` → `randmix-large-95r5w` = `rw=randrw rwmixread=95 bssplit=4k/2:64k/3:128k/5:256k/10:512k/20:768k/25:1024k/35 iodepth=32 numjobs=4 direct=1`, plus randread-large/small, randwrite-small-short, seqread-large; per-device dd seqread + fio randread under `--devices`.
- **Idle gate fail-closed:** `:57` aborts with non-zero exit when ioping/iostat exceed thresholds (defaults `--idle-latency-ms 100`, `--idle-util 85`, `:44`).
- **Programmatic logging:** JSONL `/var/log/pmss/benchmark-storage.jsonl` (`:39`); `--show-last` renderer in `storageBenchmark/report.php`.
- **Sequential only:** single `foreach` over the job list (`:109`) — no concurrent instances, randomized settings/durations, or mixed sequential-load injection. Ask #4's *orchestration* layer is genuinely unbuilt (gold-plating — no driver in ~2 years).
- **No automatic trigger:** confirmed no caller in install/update/cron. The post-install trigger is the only buildable gap.
- **Ongoing QC already passive:** the `root.cron` collectors above cover the "on-going QC" intent without synthetic load.
- **Tests:** `scripts/lib/tests/development/storageBenchmarkShowLastTest.php`, `storageBenchmarkEdgeCasesTest.php`, `storageBenchSecurityTest.php`. Gap: no dedicated regression test asserting the `--require-idle` abort-on-busy path — #656 should add one.
- **Constraints (ADRs):** trigger belongs in `update-step2` per **ADR-0006/0007** (keep `install.sh` a thin bootstrapper); dropping the remote `yabs.sh` pipe **adheres to ADR-0004/0005** (shell-command guardrails / trust boundaries).
- **Cross-issue:** GUI/long-term-result residue → **#501** (open successor); **not #45** (open but `should-not-implement`). A future guardrailed active-fio variant would depend on cgroup io.max coverage (**#273**, now closed) plus **#121/#437** — and operator re-approval.

### Recommendations
- **Implementation:** #51 as written should not be implemented (labeled `should-not-implement`); the only buildable work is tracked in **#656**.
- A guardrailed active-fio-in-production variant is a *product* decision, not an implementation task: it needs operator re-approval, verified per-user cgroup I/O limits, and evidence that passive telemetry is insufficient. It should not be reintroduced from this issue without that decision.
- Operator: #51 is fully decomposed (engine shipped; #656 carries the one buildable slice; residue → #501; active-fio + yabs refused). It can be closed as decomposed at your discretion — left open here because closure is operator-reserved.

### Worth-Implementing Analysis
- **Cost:** umbrella as written = high and partly dangerous; the surviving slice (#656) = low (~30–60 LOC reusing existing engine/gate/JSONL).
- **Benefit:** umbrella = mostly already realized or negative (ask #5 is net-harmful); #656 = real, asymmetric — catches a dud disk before a paying customer lands, at trivial cost.
- **Verdict rationale:** "Implement #51 as written" fails on every axis — redundant (engine shipped), dangerous (ask #5), or misrouted (GUI → #501). The only positive-ROI work is #656.
- **Alternative:** decompose (done) — #656 for the safe slice; refuse ask #5 as specified; drop yabs; route GUI to #501.

### Anti-Pattern Hunt
- REGEX_FOR_SEMANTICS: PASS — metrics are numeric; no semantic classification.
- PREMATURE_OPTIMIZATION: PASS — the speculative concurrent/randomized fio orchestration and the production-QC/auto-disable machinery are explicitly *not* endorsed (no driver in ~2 years); only the small, justified #656 slice survives (Exists: engine+gate+JSONL; Tested: yes; Useful: install acceptance/dud-disk catch; not-too-expensive).
- OPERATOR_HANDOFF_ANTIPATTERN: PASS — verdict is decisive with a concrete spin-out and label, not a deferral; the only operator-reserved item (closing #51, the active-fio product call) is genuinely lifecycle/authority-gated.
- LAZY_HALT_AFTER_AUTHORIZATION: PASS — terminal action taken (spin-out #656 created, label applied), not stopped at "recommend".
- FALSE_COMPLETION_CLAIM: PASS — verdict is SHOULD-NOT-IMPLEMENT (umbrella); every "shipped" claim cites a file:line read this pass; gaps enumerated.
- FABRICATED_AUTHORITY_OR_PATH: PASS — every path/line/ADR/issue cited was verified against source/API this pass.
- FAIL_OPEN_SAFETY_BEHAVIOR: PASS — `--require-idle` verified fail-closed (non-zero exit on busy); #656 enforces it plus a no-users guard.
- DRY_RUN_THEATER: PASS — real source reads, no dry-run substitution.
- INVESTIGATION_LOOP_THEATER: PASS — this is not another re-derivation: it pairs verification with the terminal decomposition action (spin-out + `should-not-implement` label) that the prior analysis recommended.
- LACK_OF_INVESTIGATION: PASS — counter-signals checked: (a) is the active-fio danger real? — verified against the in-repo fail-closed idle gate and the existence of cgroup-containment issues #273/#121/#437; (b) is the engine actually shipped? — read `storageBenchmark.php` directly rather than trusting the issue or prior comments.

### Merit Check — Tier A (umbrella as the unit of work)
1. **Concrete present demand.** Theoretical — no present demand. The issue is operator-authored from 2024-02-15 ("We should do a series of benchmarks after installation"); no current ticket, customer, or cron needs the umbrella built today. The one concrete need (install acceptance baseline) is carved into #656.
2. **Function.** As written it would (a) add a scheduled production fio caller + an auto-disable-new-users provisioning gate, (b) pipe a remote `yabs.sh`/speed-test, (c) add a result GUI, (d) add a post-install/pre-user benchmark loop. Most of (a–c) is either shipped, unsafe, or belongs to #501; only the post-install trigger is wanted.
3. **Success metric.** None definable for the umbrella (it is a 2024 wishlist, not a single change). #656's metric: at least one install run emits a peer-comparison `WARN:` to the operator log on a server with a known-slow disk, before that server takes a customer — a signal observable independent of the code's own existence.
4. **Catastrophic if implemented.** Ask #5 path: scheduled heavy fio on a populated host → host I/O/CPU saturation → soft-lockup → all co-located tenants frozen → hours of detection lag → manual power-cycle recovery. Two-plus causal steps, and it has a real precedent shape.
5. **Catastrophic if NOT implemented.** For the umbrella: nothing — the engine and passive QC already exist. The asymmetry (#4 ≫ #5) is itself the SHOULD-NOT-IMPLEMENT signal. (For #656 specifically, the "not implemented" cost is modest-but-real: dud disks surface as customer complaints rather than install-time warnings.)
6. **Deletion alternative.** Delete the umbrella from the build queue entirely; keep only #656. The engine, idle gate, JSONL, peer-comparison, and passive cron collectors already exist — nothing new to build except one guarded trigger.
7. **Cost.** (a) Build: umbrella high; #656 ~30–60 LOC. (b) Maintenance/yr: the active-fio path would require a standing "disable during incidents" runbook entry and tuning of the auto-disable threshold — recurring burden; #656 has none beyond the engine it already reuses. (c) Blast radius if it breaks: ask #5 = every tenant on a host; #656 = none (runs only on empty/idle hosts). (d) Undo cost: #656 = clean `git revert` of one caller; ask #5 path = operational incident, not a revert.
8. **Dependency blast (Hyrum).** Nothing currently depends on the umbrella (no caller of `storageBenchmark` outside its own wrapper/tests/lint). #656 adds a dependency on the `update-step2` post-provision path and on `user/selection.php`; the file is under an active decomposition refactor, so any implementation should rebase first.
9. **Requirements challenge.** The dumbest requirement is "active fio in production + auto-disable new users," because it endangers live tenants to measure a number (idle-time ceiling) that does not answer the provisioning question (contended headroom) — which passive PSI/iostat already answer.

**Tier B — Goodhart (the auto-disable gate adds a threshold/policy):** optimizing "keep the synthetic benchmark number high" incentivizes gaming via idle-time scheduling (run the QC only when the host is empty, get a great number, admit users, saturate) — i.e. the metric can be maximized without achieving the goal (real headroom under load). This is a second reason ask #5's provisioning gate is the wrong mechanism.

### Decomposition Targets
The engine was recently decomposed (the active refactor campaign) and is in good shape (~163 LOC, small focused functions). One genuine readability target remains:
- [REFACTOR] `storageBenchmark.php:96` — `fioRun()` is a single dense ~1-line function spanning option assembly, command execution, JSON parsing, and metric aggregation. Splitting "build fio argv" / "parse fio JSON → metrics" into named helpers would improve readability and testability. Low priority; not blocking.

evidence mined: searched memory (fio production saturation, post-install baseline, investigation-loop) + GH issues (`benchmark in:title`, #45/#501/#378/#121/#437/#273) -> surfaced the shipped engine + fail-closed idle gate, the real production-fio saturation hazard, the residue-routing target (#501 not #45), and the active-fio containment dependency (#273 closed) -> used all to ground the active-fio refusal, the #501 routing, and the #656 spin-out spec.

— Väinämöinen
